src/main/java/io/vertx/cassandra/impl/CassandraClientImpl.java

```diff
@@ -32,23 +32,77 @@
 import io.vertx.core.Handler;
 import io.vertx.core.Vertx;
 import io.vertx.core.impl.VertxInternal;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.shareddata.LocalMap;
+import io.vertx.core.shareddata.Shareable;
 
+import java.io.Closeable;
+import java.io.IOException;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Function;
 
 public class CassandraClientImpl implements CassandraClient {
 
+  private static final String DS_LOCAL_MAP_NAME = "__vertx.CassandraClient.datasources";
+
   VertxInternal vertx;
-  CassandraClientOptions options;
-  AtomicReference<Session> session = new AtomicReference<>(null);
+  final CassandraHolder cassandraHolder;
 
-  public CassandraClientImpl() {
+  public CassandraClientImpl(Vertx vertx, String dataSourceName, CassandraClientOptions cassandraClientOptions) {
+    this.vertx = (VertxInternal) vertx;
+    this.cassandraHolder = lookupHolder(dataSourceName, cassandraClientOptions);
   }
 
-  public CassandraClientImpl(Vertx vertx, CassandraClientOptions cassandraClientOptions) {
-    this.vertx = (VertxInternal) vertx;
-    this.options = cassandraClientOptions;
+  class CassandraHolder implements Closeable, Shareable {
+    int refCount = 1;
+    AtomicReference<Session> session = new AtomicReference<>(null);
+    CassandraClientOptions options;
+    Runnable closeRunner;
+
+    public CassandraHolder(CassandraClientOptions options, Runnable closeRunner) {
+      this.options = options;
+      this.closeRunner = closeRunner;
+    }
+
+    synchronized void incRefCount() {
+      refCount++;
+    }
+
+    @Override
+    public synchronized void close() {
+      if (--refCount == 0) {
+        if (session.get() != null) {
+          session.get().close();
+        }
+        if (closeRunner != null) {
+          closeRunner.run();
+        }
+      }
+    }
+  }
+
+  private CassandraHolder lookupHolder(String dataSourceName, CassandraClientOptions cassandraClientOptions) {
+    synchronized (vertx) {
```

Comment:
In general, you should avoid synchronizing on general objects like the Vert.x instance. If all modules do the same, it may easily create a contention point.

Reply:
Locking on a specific object would be fine (eg on the internal map instance). Synchronizing is not too bad here, since we only do it on verticle start/stop (usually). Ideally we'd have a lock free implementation since the shared clients can be used by different verticles and thus different threads.